

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>The EM2Tools API reference &mdash; EM2Tools 0.1.0 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Getting started" href="../getting_started.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> EM2Tools
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">The EM2Tools API reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-em2lib.seq_record">The “seq_record” module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-em2lib.seq">The “seq” module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-em2lib.seq_feature">The “seq_feature” module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-em2lib.seq_utils">The “seq_utils” module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-em2lib.argparse_em2">The “argparse_em2” module</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">EM2Tools</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>The EM2Tools API reference</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/api/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-em2lib">
<span id="the-em2tools-api-reference"></span><h1>The EM2Tools API reference<a class="headerlink" href="#module-em2lib" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-em2lib.seq_record">
<span id="the-seq-record-module"></span><h2>The “seq_record” module<a class="headerlink" href="#module-em2lib.seq_record" title="Permalink to this headline">¶</a></h2>
<p>Extension module to the Biopython Bio.SeqRecord module</p>
<dl class="py class">
<dt id="em2lib.seq_record.SeqRecordEM2">
<em class="property">class </em><code class="sig-prename descclassname">em2lib.seq_record.</code><code class="sig-name descname">SeqRecordEM2</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2" title="Permalink to this definition">¶</a></dt>
<dd><p>Extension to Biopython SeqRecord class</p>
<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.overlap">
<code class="sig-name descname">overlap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">strand</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves features that overlap a given position range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strand</strong> – strand specification of features to be returned. If strand is 0, then        features on both strands are returned. If feature.strand is 0, then all strands will match.</p></li>
<li><p><strong>start</strong> – start of range</p></li>
<li><p><strong>end</strong> – end of range, if None, then end=start</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of overlaping features</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.feature_after">
<code class="sig-name descname">feature_after</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span></em>, <em class="sig-param"><span class="n">strand</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">nearest</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.feature_after" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves the features immediately after (but not overlaping) the specified position, on one
strand or both. If nearest is True, then only the nearest ones are returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nearest</strong> – if True, only the nearest features are returned. This only makes sense when        strand is 0</p></li>
<li><p><strong>strand</strong> – strand specification of features to be returned. If strand is 0, then        features on both strands are returned</p></li>
<li><p><strong>position</strong> – the position</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of features after the specified position</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.feature_before">
<code class="sig-name descname">feature_before</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span></em>, <em class="sig-param"><span class="n">strand</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">nearest</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.feature_before" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves the features immediately before (but not overlaping) the specified position, on
one strand or both. If nearest is True, then only the nearest ones are returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nearest</strong> – if True, only the nearest features are returned. This only makes sense when        strand is 0</p></li>
<li><p><strong>strand</strong> – strand specification of features to be returned. If strand is 0, then        features on both strands are returned</p></li>
<li><p><strong>position</strong> – the position</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of features before the specified position</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.surrounding_features">
<code class="sig-name descname">surrounding_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span></em>, <em class="sig-param"><span class="n">strand</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">nearest</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.surrounding_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves all the features around a given position but not overlaping it. If nearest is
True, then only the nearest features are returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> – the position</p></li>
<li><p><strong>nearest</strong> – if True, only the nearest features are returned.</p></li>
<li><p><strong>strand</strong> – strand specification of features to be returned. If strand is 0, then        features on both strands are returned</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of features around the specified position</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.add_feature">
<code class="sig-name descname">add_feature</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.add_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a feature to the current record according to arguments passed as **kwargs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>kwargs</strong> – keyword arguments to pass to SeqFeatureEM2 class</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.join">
<code class="sig-name descname">join</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">offset</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">keepself</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.join" title="Permalink to this definition">¶</a></dt>
<dd><p>Joins two SeqRecordEM2 objects into a new one representing the resulting merged sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>keepself</strong> – if True and overlapping subsequences are different, then keep sequence        from self record, otherwise keep the sequence of other record.</p></li>
<li><p><strong>other</strong> – the other SeqRecordEM2 object</p></li>
<li><p><strong>offset</strong> – the offset of the two sequences. If the value is negative, then the two        sequences overlap.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the result of merging records as a new SeqRecordEM2 object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.stitch">
<code class="sig-name descname">stitch</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">fpos_in_self</span></em>, <em class="sig-param"><span class="n">fpos_in_other</span></em>, <em class="sig-param"><span class="n">feature_length</span></em>, <em class="sig-param"><span class="n">orientation</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.stitch" title="Permalink to this definition">¶</a></dt>
<dd><p>Stitches two records, that is, joins them according to an overlapping feature. The sequences
may or may not overlap. If not, Ns or Xs are added to fill the gap. If they overlap, a
warning is issued if sequences do not correspond exactly. The new record keeps track of the
two original records as Features. By convention, the self record should contain the start
position of the feature on the forward strand, the other contains the end position on the
forward strand if orientation=1 or on the reverse strand if orientation=-1. If orientation
is -1, then the other record is reversed/complemented before stitching and the position of
the overlapping feature is modified accordingly. It is the user’s responsibility to provide
the records in the right order.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – the other SeqRecordEM2 object to stitch</p></li>
<li><p><strong>fpos_in_self</strong> – feature position in self record (start position of feature)</p></li>
<li><p><strong>fpos_in_other</strong> – feature position in other record (end position of feature), according
to FeatureLocation conventions for end position requiring that length = end - start, it is
not included in the feature.</p></li>
<li><p><strong>feature_length</strong> – feature length</p></li>
<li><p><strong>orientation</strong> – the orientation of the other record relative to the self, either 1 if
it is in the same orientation, -1 if other needs to be reversed before stitching, 0 if
stranded but unknown, None for proteins</p></li>
<li><p><strong>kwargs</strong> – any additional parameters that may be passed to the constructor of the
stitching feature in the new record</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the stitched record as a new SeqRecordEM2 object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">description</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">features</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">annotations</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">letter_annotations</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">dbxrefs</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse-complement the record adjusting features and their positions accordingly. The record
id is conserved but if name is not specified ‘reversed’ is appended. All other arguments are
passed and handled by the parent method.
Note that the main goal for this method is to replace SeqRecord and Seq objects by their
SeqRecordEM2 and SeqEM2 equivalents when reverse/complementing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id</strong> – the id for the reversed record</p></li>
<li><p><strong>name</strong> – the name for the reversed record</p></li>
<li><p><strong>description</strong> – the description for the reversed record</p></li>
<li><p><strong>features</strong> – keep and adjust location of features if True</p></li>
<li><p><strong>annotations</strong> – keep annotations if True</p></li>
<li><p><strong>letter_annotations</strong> – keep letter_annotations if True</p></li>
<li><p><strong>dbxrefs</strong> – keep dbxrefs if True</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a reversed copy of the record</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_record.SeqRecordEM2.orfs_to_features">
<code class="sig-name descname">orfs_to_features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">['ATG']</span></em>, <em class="sig-param"><span class="n">stop</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">filter</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">add</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_record.SeqRecordEM2.orfs_to_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines all open reading frames in a sequence record. All the returned ORFs have a length
that is a multiple of 3. Thus, for sequences without any stop codon, 3 ORFs are returned,
one for each frame. Both strands are examined but it is possible to filter the ORFs by
length, frame, etc. with the FeatureFilter defined by the filter argument.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – a list of accepted start codons</p></li>
<li><p><strong>stop</strong> – a list of accepted stop codons</p></li>
<li><p><strong>filter</strong> – a FeatureFilter object defining a filter to select ORFs according to some
criteria</p></li>
<li><p><strong>add</strong> – if True, the selected ORFs are added to the record’s features</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of ORFs as SeqFeatureEM2 objects</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-em2lib.seq">
<span id="the-seq-module"></span><h2>The “seq” module<a class="headerlink" href="#module-em2lib.seq" title="Permalink to this headline">¶</a></h2>
<p>Extension of Bio.Seq.Seq class from Biopython to add or improve functionalities</p>
<dl class="py class">
<dt id="em2lib.seq.SeqEM2">
<em class="property">class </em><code class="sig-prename descclassname">em2lib.seq.</code><code class="sig-name descname">SeqEM2</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em>, <em class="sig-param"><span class="n">seqtype</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq.SeqEM2" title="Permalink to this definition">¶</a></dt>
<dd><p>SeqEM2 class providing extension to Bio.Seq.Seq class of BioPython package.</p>
<dl class="py method">
<dt id="em2lib.seq.SeqEM2.dna">
<em class="property">classmethod </em><code class="sig-name descname">dna</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq.SeqEM2.dna" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a DNA sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – the sequence string</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a SeqEM2 DNA instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq.SeqEM2.protein">
<em class="property">classmethod </em><code class="sig-name descname">protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq.SeqEM2.protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a protein sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> – the sequence string</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a SeqEM2 protein instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq.SeqEM2.is_protein">
<code class="sig-name descname">is_protein</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq.SeqEM2.is_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Tests whether sequence was created as a protein</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>boolean, True if sequence was created as a protein, False otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq.SeqEM2.length_in_range">
<code class="sig-name descname">length_in_range</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">minlength</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">maxlength</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq.SeqEM2.length_in_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks whether the sequence length is with the specified range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>minlength</strong> – lower length bound</p></li>
<li><p><strong>maxlength</strong> – upper length bound</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if sequence length is within specified range, False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq.SeqEM2.re_search">
<code class="sig-name descname">re_search</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">regex</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq.SeqEM2.re_search" title="Permalink to this definition">¶</a></dt>
<dd><p>Searches a sequence using a regular expression</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regex</strong> – the regular expression</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of re.Match instances</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq.SeqEM2.search">
<code class="sig-name descname">search</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pattern</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq.SeqEM2.search" title="Permalink to this definition">¶</a></dt>
<dd><p>Searches sequence for a pattern specified with a fuzznuc or fuzzpro like syntax</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pattern</strong> – the pattern to be searched for that is converted into a regular expression</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of re.Match objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq.SeqEM2.get_orfs">
<code class="sig-name descname">get_orfs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">['ATG']</span></em>, <em class="sig-param"><span class="n">stop</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq.SeqEM2.get_orfs" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines all open reading frames in a sequence. It only examines the forward strand. All
the returned ORFs have a length that is a multiple of 3. Thus, for sequences without any
stop codon, 3 ORFs are returned, one for each frame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – list of accepted start codon or None if ORFs do not need to start at a start
codon.</p></li>
<li><p><strong>stop</strong> – list of accepted stop codons</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a set of tuples (start, end) of orfs where start is the starting positon of the
orf and end its ending position, not including the stop codon</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-em2lib.seq_feature">
<span id="the-seq-feature-module"></span><h2>The “seq_feature” module<a class="headerlink" href="#module-em2lib.seq_feature" title="Permalink to this headline">¶</a></h2>
<p>Extension of Bio.SeqFeature module from Biopython to add or improve functionalities</p>
<dl class="py class">
<dt id="em2lib.seq_feature.SeqFeatureEM2">
<em class="property">class </em><code class="sig-prename descclassname">em2lib.seq_feature.</code><code class="sig-name descname">SeqFeatureEM2</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">parent</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.SeqFeatureEM2" title="Permalink to this definition">¶</a></dt>
<dd><p>SeqFeatureEM2 class providing extension to Bio.SeqFeature.SeqFeature class of BioPython package.</p>
<dl class="py method">
<dt id="em2lib.seq_feature.SeqFeatureEM2.length_in_range">
<code class="sig-name descname">length_in_range</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">minlength</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">maxlength</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.SeqFeatureEM2.length_in_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks whether the feature length is with the specified range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>minlength</strong> – lower length bound</p></li>
<li><p><strong>maxlength</strong> – upper length bound</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if feature length is within specified range, False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.SeqFeatureEM2.lies_within">
<code class="sig-name descname">lies_within</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.SeqFeatureEM2.lies_within" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines whether feature lies entirely with the specified range. Fuzzy positions are turned into integers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – start of range either int or ExactPosition</p></li>
<li><p><strong>end</strong> – end of range either int or ExactPosition</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if feature boundaries lie with the specified range.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.SeqFeatureEM2.covers">
<code class="sig-name descname">covers</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.SeqFeatureEM2.covers" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines whether feature covers the whole range specified by start and end</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – start of range either int or ExactPosition</p></li>
<li><p><strong>end</strong> – end of range either int or ExactPosition, if None then end=start</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if feature covers the specified range</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.SeqFeatureEM2.overlaps">
<code class="sig-name descname">overlaps</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.SeqFeatureEM2.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines whether feature overlaps a position range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – start of range either int or ExactPosition</p></li>
<li><p><strong>end</strong> – end of range either int or ExactPosition</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if feature overlaps range</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.SeqFeatureEM2.intersect">
<code class="sig-name descname">intersect</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.SeqFeatureEM2.intersect" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a new feature which is the intersection of feature and another one</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> – the other feature</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.SeqFeatureEM2.move">
<code class="sig-name descname">move</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">offset</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.SeqFeatureEM2.move" title="Permalink to this definition">¶</a></dt>
<dd><p>Moves a feature by a certain offset</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>offset</strong> – offset by which the feature must be moved</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="em2lib.seq_feature.FeatureFilter">
<em class="property">class </em><code class="sig-prename descclassname">em2lib.seq_feature.</code><code class="sig-name descname">FeatureFilter</code><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>A class for the definition and application of a filter to a list of features. May be used to
retrieve features from a record or any list of features according to length, location, type,
strand and frame. It is possible to combine several criteria into one single filter.</p>
<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.keep">
<code class="sig-name descname">keep</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.keep" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the keep attribute of the filter. If True, the features corresponding to the specified
criteria will be kept, otherwise, they will be discarded.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>keep</strong> – boolean, True if features consistent with the criteria should be kept.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the current filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.type">
<code class="sig-name descname">type</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feat_type</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.type" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the type of feature.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feat_type</strong> – the type of feature to select</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the current filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.length">
<code class="sig-name descname">length</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">minlength</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">maxlength</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.length" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the length range of the feature.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>minlength</strong> – the minimum length of the feature or 0 if None</p></li>
<li><p><strong>maxlength</strong> – the maximum length of the feature or no limit if set to None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the current filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.strand">
<code class="sig-name descname">strand</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strand</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the strand of the feature.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>strand</strong> – the strand of the feature</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the current filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.frame">
<code class="sig-name descname">frame</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">frame</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">strand</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.frame" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the frame of the filter. By default, this is on the forward strand unless the strand is set to -1.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frame</strong> – the frame 0, 1 or 2</p></li>
<li><p><strong>strand</strong> – the strand 1 or - 1</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the current filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.covers">
<code class="sig-name descname">covers</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.covers" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a region that must be covered by the feature.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – start position of the region</p></li>
<li><p><strong>end</strong> – end position of the region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the current filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.overlaps">
<code class="sig-name descname">overlaps</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a region that must overlap the feature.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – start position of the region</p></li>
<li><p><strong>end</strong> – end position of the region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the current filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.lies_within">
<code class="sig-name descname">lies_within</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.lies_within" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a region within which the feature must lie.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – start position of the region</p></li>
<li><p><strong>end</strong> – end position of the region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the current filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.type_applies">
<code class="sig-name descname">type_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.type_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if type criterion applies to the feature and return a boolean stating whether the
feature should be kept or rejected. If the criterion has not been set, then True is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True the feature must be kept</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.length_applies">
<code class="sig-name descname">length_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.length_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if length criterion applies to the feature and return a boolean stating whether the
feature should be kept or rejected. If the criterion has not been set, then True is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True the feature must be kept</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.covers_applies">
<code class="sig-name descname">covers_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.covers_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if covers criterion applies to the feature and return a boolean stating whether the
feature should be kept or rejected. If the criterion has not been set, then True is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True the feature must be kept</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.overlaps_applies">
<code class="sig-name descname">overlaps_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.overlaps_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if overlaps criterion applies to the feature and return a boolean stating whether the
feature should be kept or rejected. If the criterion has not been set, then True is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True the feature must be kept</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.lies_within_applies">
<code class="sig-name descname">lies_within_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.lies_within_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if lies_within criterion applies to the feature and return a boolean stating
the feature should be kept or rejected. If the criterion has not been set, then True is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True the feature must be kept</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.location_applies">
<code class="sig-name descname">location_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.location_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if location criterion applies to the feature and return a boolean stating whether the
feature should be kept or rejected. If the criterion has not been set, then True is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True the feature must be kept</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.strand_applies">
<code class="sig-name descname">strand_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.strand_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if strand criterion applies to the feature and return a boolean stating whether the
feature should be kept or rejected. If the criterion has not been set, then True is
returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True the feature must be kept</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.frame_applies">
<code class="sig-name descname">frame_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.frame_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if frame criterion applies to the feature and return a boolean stating whether the
feature should be kept or rejected. If the criterion has not been set, then True is
returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True the feature must be kept</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_feature.FeatureFilter.apply">
<code class="sig-name descname">apply</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">features</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_feature.FeatureFilter.apply" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if all defined criteria apply to the features in the list and return the list of
features corresponding to the specified criteria.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>features</strong> – the list of features to filter</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the filtered list of features</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-em2lib.seq_utils">
<span id="the-seq-utils-module"></span><h2>The “seq_utils” module<a class="headerlink" href="#module-em2lib.seq_utils" title="Permalink to this headline">¶</a></h2>
<p>Extension module to the Biopython Bio.SeqUtils module</p>
<dl class="py function">
<dt id="em2lib.seq_utils.ambiguous2string">
<code class="sig-prename descclassname">em2lib.seq_utils.</code><code class="sig-name descname">ambiguous2string</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">code</span></em>, <em class="sig-param"><span class="n">protein</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.ambiguous2string" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an ambiguous residue into a string with all compatible unambiguous residues.
If the input code is not ambiguous, it is returned without any conversion.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>code</strong> – the input code to be converted into a list of residues.</p></li>
<li><p><strong>protein</strong> – True if residue is amino-acid</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a string corresponding to the unambiguous residues compatible with the input code</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="em2lib.seq_utils.isambiguous">
<code class="sig-prename descclassname">em2lib.seq_utils.</code><code class="sig-name descname">isambiguous</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">code</span></em>, <em class="sig-param"><span class="n">protein</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.isambiguous" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks code is an ambiguous residue specification or not.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>code</strong> – the input code that must be checked for ambiguity</p></li>
<li><p><strong>protein</strong> – True if code is amino-acid code</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>boolean, True if code is ambiguous, False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="em2lib.seq_utils.pattern2regex">
<code class="sig-prename descclassname">em2lib.seq_utils.</code><code class="sig-name descname">pattern2regex</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pattern</span></em>, <em class="sig-param"><span class="n">protein</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.pattern2regex" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a fuzznuc or fuzzpro-like pattern into a regular expression that can be used to search
a sequence string.</p>
<ul class="simple">
<li><p>[ABC] =&gt; any of ABC residues,</p></li>
<li><p>{ABC} =&gt; any residue except ABC,</p></li>
<li><p>&lt;ABC… =&gt; start of sequence,</p></li>
<li><p>…ABC&gt; =&gt; end of sequence,</p></li>
<li><p>A(n)(ABC)(n) =&gt; repeat residue or subsequence n times,</p></li>
<li><p>A(n,m)(ABC)(n,m) =&gt; repeat residue or subsequence from n up to m times.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pattern</strong> – the pattern definition (string)</p></li>
<li><p><strong>protein</strong> – True if pattern applies to a protein sequence, False otherwise.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the regular expression pattern as a string</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="em2lib.seq_utils.SeqFilter">
<em class="property">class </em><code class="sig-prename descclassname">em2lib.seq_utils.</code><code class="sig-name descname">SeqFilter</code><a class="headerlink" href="#em2lib.seq_utils.SeqFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>A class for the creation of a sequence filter to specify filtering criteria and applying the     filter to a list of sequence records.</p>
<dl class="py method">
<dt id="em2lib.seq_utils.SeqFilter.length">
<code class="sig-name descname">length</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">minlength</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">maxlength</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.SeqFilter.length" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimal and maximal length specification</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>minlength</strong> – minimal accepted length</p></li>
<li><p><strong>maxlength</strong> – maximal accepted length</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>SeqFilter instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.SeqFilter.pattern">
<code class="sig-name descname">pattern</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pattern</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.SeqFilter.pattern" title="Permalink to this definition">¶</a></dt>
<dd><p>pattern specification</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pattern</strong> – pattern that must be in the sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>SeqFilter instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.SeqFilter.name">
<code class="sig-name descname">name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.SeqFilter.name" title="Permalink to this definition">¶</a></dt>
<dd><p>sequence record name specification</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>name</strong> – name regular expression</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>SeqFilter instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.SeqFilter.keep">
<code class="sig-name descname">keep</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">keep</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.SeqFilter.keep" title="Permalink to this definition">¶</a></dt>
<dd><p>Boolean defining whether the matching sequences must be kept (True) or removed (False)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>keep</strong> – True to keep positive sequences, False to remove them</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>SeqFilter instance</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.SeqFilter.length_applies">
<code class="sig-name descname">length_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rec</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.SeqFilter.length_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>test whether length criterion applies to the sequence record</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>rec</strong> – the sequence record to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>boolean True if criterion applies or False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.SeqFilter.pattern_applies">
<code class="sig-name descname">pattern_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rec</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.SeqFilter.pattern_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>test whether parameter criterion applies to the sequence record</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>rec</strong> – the sequence record to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>boolean True if criterion applies or False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.SeqFilter.name_applies">
<code class="sig-name descname">name_applies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rec</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.SeqFilter.name_applies" title="Permalink to this definition">¶</a></dt>
<dd><p>test whether name criterion applies to the sequence record</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>rec</strong> – the sequence record to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>boolean True if criterion applies or False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.SeqFilter.apply">
<code class="sig-name descname">apply</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">records</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.SeqFilter.apply" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters a list of SeqRecords instances, keeping only records satisfying the specified
criteria of length, match of a pattern, name specification. It is possible to invert
the filtering process by setting the keep boolean to False and thus only keep records
which do not satisfy the criteria.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>records</strong> – list of SeqRecord instances to apply</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the filtered list of records</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="em2lib.seq_utils.GFF">
<em class="property">class </em><code class="sig-prename descclassname">em2lib.seq_utils.</code><code class="sig-name descname">GFF</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature_list</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">input_df</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.GFF" title="Permalink to this definition">¶</a></dt>
<dd><p>Manipulation of features based upon gffpandas package</p>
<dl class="py method">
<dt id="em2lib.seq_utils.GFF.add_feature_list">
<code class="sig-name descname">add_feature_list</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature_list</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.GFF.add_feature_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a list of feature to the list of an existing GFF object</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature_list</strong> – list of features to add to DataFrame</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the GFF object with feature list appended</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.GFF.df_from_feature">
<em class="property">static </em><code class="sig-name descname">df_from_feature</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">feature</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.GFF.df_from_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a pandas DataFrame from a feature (SeqFeatureEM2 or SeqFeature)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature</strong> – the feature to convert into a dataframe</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the resulting dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="em2lib.seq_utils.GFF.to_feature_list">
<code class="sig-name descname">to_feature_list</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">parents</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.seq_utils.GFF.to_feature_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts features in a GFF object into a list of SeqFeatureEM2 objects</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>parents</strong> – list of references to parent SeqRecord objects or a single parent reference
if all features are defined in the same parent. If it is a list, it should be of the same
length as the dataframe, repeating references as needed to get the right number.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of SeqFeatureEM2 objects</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-em2lib.argparse_em2">
<span id="the-argparse-em2-module"></span><h2>The “argparse_em2” module<a class="headerlink" href="#module-em2lib.argparse_em2" title="Permalink to this headline">¶</a></h2>
<p>Extension module to the standard argparse module. Adding custom actions and argument
verification methods.</p>
<dl class="py class">
<dt id="em2lib.argparse_em2.GetList">
<em class="property">class </em><code class="sig-prename descclassname">em2lib.argparse_em2.</code><code class="sig-name descname">GetList</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">option_strings</span></em>, <em class="sig-param"><span class="n">dest</span></em>, <em class="sig-param"><span class="n">nargs</span><span class="o">=</span><span class="default_value">'+'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.argparse_em2.GetList" title="Permalink to this definition">¶</a></dt>
<dd><p>An argparse custom action to return a list from an argument containing a list of elements
and/or file names. Files are supposed to contain one element of the list per line. There
can be more than one file and the argument may take a combination of elements and files.
In all cases, the returned list will contain all the specified elements without any
checking for redundancy. If you need a non redundant set instead of a list, then use
GetSet action instead.</p>
<dl class="py method">
<dt id="em2lib.argparse_em2.GetList.arg2list">
<em class="property">static </em><code class="sig-name descname">arg2list</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">values</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.argparse_em2.GetList.arg2list" title="Permalink to this definition">¶</a></dt>
<dd><p>This method converts the argument values containing elements and/or files containing
elements into a list of elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>values</strong> – argument values, this is supposed to be a list of arguments or None        (returns an empty list)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the list of elements or an empty list if the argument was None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="em2lib.argparse_em2.GetSet">
<em class="property">class </em><code class="sig-prename descclassname">em2lib.argparse_em2.</code><code class="sig-name descname">GetSet</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">option_strings</span></em>, <em class="sig-param"><span class="n">dest</span></em>, <em class="sig-param"><span class="n">nargs</span><span class="o">=</span><span class="default_value">'+'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.argparse_em2.GetSet" title="Permalink to this definition">¶</a></dt>
<dd><p>An argparse custom action to return a set from an argument containing a list of elements
and/or file names. Files are supposed to contain one element of the list per line. There
can be more than one file and the argument may take a combination of elements and files.
In all cases, the returned set will contain all the specified elements keeping only
one copy of each element. If you do not want to remove redundancy, then use GetList
action instead.</p>
<dl class="py method">
<dt id="em2lib.argparse_em2.GetSet.arg2set">
<em class="property">static </em><code class="sig-name descname">arg2set</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">values</span></em><span class="sig-paren">)</span><a class="headerlink" href="#em2lib.argparse_em2.GetSet.arg2set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method converts the argument values containing elements and/or files containing
elements into a set of elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>values</strong> – argument values, this is supposed to be a list of arguments or None        (returns an empty set)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the set of elements or an empty set if the argument was None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../getting_started.html" class="btn btn-neutral float-left" title="Getting started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Frédéric PLEWNIAK

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>